src/Adapter/Carrier/Repository/CarrierRepository.php

```diff
@@ -69,73 +79,133 @@ public function add(Carrier $carrier): CarrierId
+            ->setParameter('carrierId', $carrierId->getValue())
+        ;
+
+        return array_map(static function (array $shop) {
+            return new ShopId((int) $shop['id_shop']);
+        }, $qb->executeQuery()->fetchAllAssociative());
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     * @param ShopGroupId $shopGroupId
+     *
+     * @return ShopId[]
+     */
+    public function getAssociatedShopIdsFromGroup(CarrierId $carrierId, ShopGroupId $shopGroupId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('ps.id_shop')
+            ->from($this->dbPrefix . 'carrier_shop', 'ps')
+            ->innerJoin(
+                'ps',
+                $this->dbPrefix . 'shop',
+                's',
+                's.id_shop = ps.id_shop'
+            )
+            ->where('ps.id_carrier = :carrierId')
+            ->andWhere('s.id_shop_group = :shopGroupId')
+            ->setParameter('shopGroupId', $shopGroupId->getValue())
+            ->setParameter('carrierId', $carrierId->getValue())
+        ;
+
+        return array_map(static function (array $shop) {
+            return new ShopId((int) $shop['id_shop']);
+        }, $qb->executeQuery()->fetchAllAssociative());
+    }
+
+    /**
+     * Create a new version of the carrier, or return the carrier as is if it don't have order linked.
+     *
+     * @param Carrier $carrier
+     *
+     * @return Carrier
+     */
+    public function createNewVersion(Carrier $carrier): Carrier
+    {
+        $carrierId = new CarrierId((int) $carrier->id);
+
+        // If the carrier don't have orders linked, we can return it as is
+        if (!$this->carrierHasOrders($carrierId)) {
+            return $carrier;
         }
 
-        $newCarrier->deleted = false; // just to be sure...
+        // Otherwise, we need to create a new version of the carrier
+        /** @var Carrier $newCarrier */
+        $newCarrier = $carrier->duplicateObject();
+        $carrier->deleted = true;
+        $this->partiallyUpdateObjectModel($carrier, ['deleted'], CannotUpdateCarrierException::class);
 
         // Copy all others information like ranges, shops associated, ...
         $newCarrier->copyCarrierData($carrierId->getValue());
-
         $this->updateObjectModel($newCarrier, CannotUpdateCarrierException::class);
+        $newCarrier->setGroups($carrier->getAssociatedGroupIds());
 
-        $newCarrier->setGroups($oldCarrier->getAssociatedGroupIds());
-
+        // Return the new duplicated carrier
         return $newCarrier;
     }
+
+    public function carrierHasOrders(CarrierId $carrierId): bool
+    {
+        $qb = $this->connection->createQueryBuilder();
+
+        $count = $qb->select('COUNT(*)')
+            ->from($this->dbPrefix . 'order_carrier', 'oc')
+            ->where('oc.id_carrier = :carrierId')
+            ->setParameter('carrierId', $carrierId->getValue())
+            ->groupBy('oc.id_carrier')
```

Comment:
I think the group by is not necessary, no ?

Reply:
Oh yes, you're right! Thanks! ;)